presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -181,7 +182,15 @@ public ActualProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<Ac
         @Override
         public ActualProperties visitAssignUniqueId(AssignUniqueId node, List<ActualProperties> inputProperties)
         {
-            return Iterables.getOnlyElement(inputProperties);
+            ActualProperties properties = Iterables.getOnlyElement(inputProperties);
+            if (properties.getNodePartitioning().isPresent()) {
+                // preserve input (possibly preferred) partitioning
+                return properties;
+            }
+
+            return ActualProperties.builderFrom(properties)
+                    .global(partitionedOn(FIXED_HASH_DISTRIBUTION, node.getOutputSymbols(), Optional.empty()))
```

Comment:
Why FIXED_HASH_DISTRIBUTION? 

The addition of `unique` property ensures that the data is partitioned on all columns, but there are no guarantees about how it is partitioned.

Also, I think `partitioned_on(<all-columns>)` is weaker than `partitioned_on(unique)`. `partitioned_on(unique)` implies partitioning on unique+<any combination of columns>`, e.g. partitioned_in(unique, col1) as well as partitioned_in(unique, col1, col2), and enables one-step aggregation for different grouping sets, e.g. `GROUP BY unique, col1` as well as `GROUP BY unique col1, col2`.

Reply:
> Why FIXED_HASH_DISTRIBUTION?

Yeah. I am not sure why `Partitioning` (with `PartitioningHandle`) is part of `ActualProperties` (it seems to be unused). I guess it's for convenience. We could add something like `SOURCE_HASH_DISTRIBUTION` or try to untangle `Partitioning` from actual properties.

> Also, I think partitioned_on(<all-columns>) is weaker than partitioned_on(unique). partitioned_on(unique) implies partitioning on unique+, e.g. partitioned_in(unique, col1) as well as partitioned_in(unique, col1, col2), and enables one-step aggregation for different grouping sets, e.g.GROUP BY unique, col1as well asGROUP BY unique col1, col2`.

True.

Reply:
> Yeah. I am not sure why Partitioning (with PartitioningHandle) is part of ActualProperties (it seems to be unused). I guess it's for convenience

So I think it's because one might want to insert remote exchange for one source (e.g: of a join) to match partitioning of other source (see: com/facebook/presto/sql/planner/optimizations/AddExchanges.java:739)

However, this should be correct in this case as rows are node partitioned on `unique` and there is fixed number of nodes, so technically `FIXED_HASH_DISTRIBUTION` applies. In practice, `unique` won't be ever matcher as part of any key, so any system hash partitioning handle will do

Reply:
@sopel39 Karol, 

> However, FIXED_HASH_DISTRIBUTION should be correct in this case as rows are node partitioned on unique and there is fixed number of nodes, so technically FIXED_HASH_DISTRIBUTION applies. In practice, unique won't be ever matched as part of any key (join or union), so any system hash partitioning handle will do

It would be cleaner to extend `ActualProperties#Global` to allow describing `partitioned_on(columns)` without clarifying what partitioning function is. When `AssignUniqueId` is applies on top of `TableScanNode`, the partitioning is not *hash* and the number of nodes can change if new nodes get adding during query execution.